pkg/controller/node/nodecontroller.go

```diff
@@ -555,6 +565,85 @@ func (nc *NodeController) Run(stopCh <-chan struct{}) {
+func (nc *NodeController) addOrRemoveNoScheduleTaintByCondition(node *v1.Node) (bool, bool) {
+	var changedToUnSchedulable, changedToSchedulable bool
+	observedMap := map[v1.NodeConditionType]*v1.NodeCondition{}
+	_, observedMap[v1.NodeReady] = v1.GetNodeCondition(&node.Status, v1.NodeReady)
+	_, observedMap[v1.NodeOutOfDisk] = v1.GetNodeCondition(&node.Status, v1.NodeOutOfDisk)
+	_, observedMap[v1.NodeNetworkUnavailable] = v1.GetNodeCondition(&node.Status, v1.NodeNetworkUnavailable)
+
+	// If there are saved node status, it only make sense to check conditions when some change happened.
+	if savedNodeStatus, found := nc.nodeStatusMap[node.GetName()]; found {
+		for conditionType, condition := range observedMap {
+			_, savedCondition := v1.GetNodeCondition(&savedNodeStatus.status, conditionType)
+			// something changed
+			if (condition != nil && savedCondition == nil) ||
+				(condition != nil && savedCondition != nil && condition.Status != savedCondition.Status) {
+				changedToSchedulable, changedToUnSchedulable = setSchedulingFlag(condition)
+				if changedToUnSchedulable {
+					break
+				}
+			}
+		}
+	} else {
+		// Otherwise, there are no saved node status, it always worth to check conditions
+		for _, condition := range observedMap {
+			if condition != nil {
+				changedToSchedulable, changedToUnSchedulable = setSchedulingFlag(condition)
+				if changedToUnSchedulable {
+					break
+				}
+			}
+		}
+	}
+
+	if changedToUnSchedulable {
+		if err := controller.AddOrUpdateTaintOnNode(nc.kubeClient, node.Name, NoScheduleTaintTemplate); err != nil {
+			glog.Errorf("failed to add NoSchedule taint on node %v, %v", node.GetName(), err)
+		} else {
+			glog.V(2).Infof("Tainting Node %v with NoSchedule taint because its condition changed to unschedulable",
+				node.Name,
+			)
+		}
+	}
+
+	if changedToSchedulable {
+		if err := controller.RemoveTaintOffNode(nc.kubeClient, node.Name, NoScheduleTaintTemplate, node); err != nil {
+			glog.Errorf("failed to remove NoSchedule taint on node %v, %v", node.GetName(), err)
+		} else {
+			glog.V(2).Infof("Mark node %v as schedulable again because its condition changed to schedulable",
+				node.Name,
+			)
+		}
+	}
+
+	return changedToSchedulable, changedToUnSchedulable
+}
+
+// setSchedulingFlag calculates the result of scheduling flag by condition
+// 1. changedToSchedulable
+// 2. changedToUnSchedulable
+func setSchedulingFlag(condition *v1.NodeCondition) (bool, bool) {
+	if !isSchedulable(condition) {
+		return false, true
+	} else {
+		return true, false
+	}
+}
+
+// isSchedulable is a flight check for node readiness, disk, and network.
+func isSchedulable(cond *v1.NodeCondition) bool {
+	if (cond.Type == v1.NodeReady) && (cond.Status != v1.ConditionTrue) {
+		return false
+	} else if (cond.Type == v1.NodeOutOfDisk) && (cond.Status != v1.ConditionFalse) {
+		return false
+	} else if (cond.Type == v1.NodeNetworkUnavailable) && (cond.Status != v1.ConditionFalse) {
```

Comment:
regarding #45717 , prefer to return different `Taint` for different condition instead of true or false.